Add common actions with shortcuts to editor background

neilcsmith-net opened the pull request:
This is an alternative implementation to #5591 to show common actions and shortcuts on the empty editor area.  This addresses concerns raised there, including added dependencies, and uses the layer system so that the options aren't hardcoded and platform applications can choose whether to add their own actions or not.  The actions here are also clickable buttons as well as shortcut displayers.

This PR also fixes issues with the background image displayer (the image component is never removed so sometimes shows over editors).  And also adds a layer attribute to configure that as well as the existing system property, as well as the ability to show actions and an image.

An image and actions have been added in to the NetBeans branding module.  (I was going to use the svg icon from NBPackage, but our Batik support doesn't like it, so it's a png for now).

![Screenshot from 2024-03-18 18-11-39](https://github.com/apache/netbeans/assets/3975960/fa0c30af-ed26-4afb-9ed1-2166d6f11502)

NB. I've added the action to open the Start Page here, as I'm also working on an idea to replace that with a dashboard API for the IDE and platform.  I was originally going to try adding that in to the editor background, but this felt cleaner, and useful by itself.

matthiasblaesing commented:
Only eyeballed this, but the decoupling looks like what I had in mind. This approach is more focused than what I thought about, but also looks sane.

I noticed that you invented your own stringification algorithm for the keystrokes. I assume, that you don't use `Utilities.keyToString` directly because you don't agree with the stringification it creates?

A question that came to my mind: how does the list react when it becomes to large for the window? Reason could be because the user resized it so small or because the list is just long.

mbien commented:
looks clean and has potential for more!

> Utilities.keyToString

Interesting - I didn't know that this existed, I actually searched for this yesterday since I wanted to add hotkeys to the tooltip of some toolbar buttons for #7128. Refactoring, quick search UI etc simply puts the default hotkeys into the bundle text - which isn't ideal.

neilcsmith-net commented:
> Only eyeballed this, but the decoupling looks like what I had in mind. This approach is more focused than what I thought about, but also looks sane.

Thanks.  I realised if we went for registering background components, we'd still need to also register the actions for that component somewhere.  There are benefits to having action display directly in the module, but that doesn't mean we couldn't extend to support other components later too.
 
> I noticed that you invented your own stringification algorithm for the keystrokes. I assume, that you don't use `Utilities.keyToString` directly because you don't agree with the stringification it creates?

Actually I copied that from KeyStrokeUtils in options.keymap (as commented).  `Utilities::keyToString` isn't really user friendly for this - eg. `DS-O` vs `Ctrl+Shift+O`.  I was thinking that we needed a utility API that provides the text you'd see in the menu, etc.  Just having another look, I realised that's already available in `Actions::keyStrokeToString`.  I've updated the PR to use that instead.

> A question that came to my mind: how does the list react when it becomes to large for the window? Reason could be because the user resized it so small or because the list is just long.

The image disappears first, then the list will just get hidden.  I thought about limiting the number of actions, but I'm not sure that's needed as and until we provide a UI to customise.  The window needs to be small enough to be unusable for the six registered here not to be in view.

Chris2011 commented:
@neilcsmith-net awesome 😎 . Thx for the nice design. One thing is that I prefer to show a text for actiosn that are not have a shortcut yet.

neilcsmith-net commented:
@Chris2011 I did try this, but I thought it looked a bit busy and distracted from the actual shortcuts.  On the other hand, it balances better.

eg.
![Screenshot from 2024-03-19 17-56-56](https://github.com/apache/netbeans/assets/3975960/8b54c71e-2649-4dc9-b088-9b8085f527fa)

Also tried a hyphen -

![Screenshot from 2024-03-19 17-56-06](https://github.com/apache/netbeans/assets/3975960/2e393caa-c559-41cb-85c2-df406890f27a)

I'm inclined to leave blank but happy to relook if others also in favour of some text.

Chris2011 commented:
You are right, no shortcut assigned doesn't fit here well with the new layout. Maybe just n/a will also be fine?

mbien commented:
I like it. In fact I like it so much that I think we should consider to turn the auto-open of the start page off by default - given that there is a button for it, where the user can open it and enable the auto-open again. (which would be weird timing since I just fixed that page https://github.com/apache/netbeans/pull/7141)

The nb logo should have a purpose too! Could it open the about window?

mbien commented on platform/core.windows/src/org/netbeans/core/windows/view/EditorView.java:
```diff
@@ -257,6 +250,64 @@ public void dropActionChanged(DropTargetDragEvent dtde) {
+                    imageComponent = label;
+                } else {
+                    Logger.getLogger(EditorView.class.getName()).log(Level.WARNING, null,
+                            new java.lang.NullPointerException("Image not found at "
```
nitpick: `java.lang.` is redundant here

neilcsmith-net replied:
Well, not my code, this was moved from elsewhere in the file.  But have removed the redundant text in latest update.

mbien reviewed (approved):
code looks good too!

what do you think about adding a `Project Groups...` button below `Open Project...`? Project groups (and the multi-project nature) are a core feature of NB at this point - not everyone who is checking NB out might know about it. Button might be even useful for veterans.

neilcsmith-net commented on platform/core.windows/src/org/netbeans/core/windows/view/EditorView.java:
```diff
@@ -257,6 +250,64 @@ public void dropActionChanged(DropTargetDragEvent dtde) {
+        }
+
+        private JComponent initBackgroundActions() {
+            java.util.List<? extends Action> actions
```
Have left for now.  That whole module could do with cleaning up based on that.

neilcsmith-net commented:
Updated the code based on comments.

> no shortcut assigned doesn't fit here well with the new layout. Maybe just n/a will also be fine?

I'm not sure anything with letters works very well.  Anyway, I've updated to make the no-shortcut label brandable.  Empty by default in the platform, set to `-` currently in the IDE.

I've used the bundle from the ui package for this, because if we extend this further the background component should probably move into there.

> In fact I like it so much that I think we should consider to switch the auto-open of the start page off by default 

Added extra commit to do that.  As said in the main description, I've started looking at some sort of dashboard API for the platform which has overlap with the welcome page.  The welcome page is showing its age and a lot of the content isn't that useful any more.

> The nb logo should have a purpose too! Could it open the about window?

We could consider extending to do that later.  Can we leave until NB23+?  I've reused (and added a fix) to the existing image support in this.  I don't know if anyone is using the `netbeans.winsys.imageSource` property as specified in https://bits.netbeans.org/dev/javadoc/org-openide-windows/overview-summary.html#group-systemproperty  

> what do you think about adding a Project Groups... button below Open Project...?

Was top of my list of actions I left out.  Came very close to adding it.  Possibly too advanced for the beginner, and not the most useful place for a regular NetBeans user?

mbien commented:
> Was top of my list of actions I left out. Came very close to adding it. Possibly too advanced for the beginner, and not the most useful place for a regular NetBeans user?

you are probably right. Once a user starts using PGs its likely that he/she won't see that panel very often unless maybe when switching to the `none` group.

regarding usefulness for beginners: well a beginner could also be someone who knows many IDEs already but checked out NB for the first time.

Chris2011 commented:
I also see the most common actions as a priority here. I also find it useful to see that we can drag a file or folder to the empty place to open this folder as a "project" but maybe also not that necessary. To add. Maybe also the quick search would be good to catch. Maybe later if we refactor this also :D.

I would suggest to have it like this, all other stuff we should add to a redesigned welcome page.

neilcsmith-net commented:
> I would suggest to have it like this, all other stuff we should add to a redesigned welcome page.

My goal with this and any dashboard API would be to make these elements foreground common tasks for beginners, seasoned users and people who have experience of other IDEs.

Both should eventually be customizable via UI, and usable in platform applications too (which is part of my reason for working on it at the moment).

A Project Group widget on a dashboard might make more sense than an action here.

The Quick Search action is the only action with a shortcut already on display, and invoking the action directly by clicking is not that useful.  Labels for eg. drag-n-drop need a little thought with this implementation.  I wonder if there's an easily understandable icon for that - could add below logo?

mhalachev commented:
The idea to add common actions with shortcuts to the editor background is quite convenient. Additionally, would it be feasible to incorporate action icons (from menus/toolbars), as this may provide added benefits for beginners? Here is a mockup what it might look like:

<img width="385" alt="mockup" src="https://github.com/apache/netbeans/assets/2463597/bc88e2f7-a17e-4aa1-a357-08ab9abff065">

Considering that this is the editor background, to enhance its lightweight appearance, while ensuring visual distinction from the toolbar, icons may be semi-transparent.

Another consideration, albeit it may be deemed excessive, is the inclusion of recent files and projects. This would facilitate quick and easy access, as users presently must navigate menus or open the Welcome screen and select the appropriate tab.

<img width="454" alt="recent" src="https://github.com/apache/netbeans/assets/2463597/ead89fcd-48a4-4edd-bc37-a1df183c840e">

If ever implemented, it could potentially display the latest 3-4 items, whereas the menus and Welcome screen may accommodate a greater number of items.

neilcsmith-net commented:
@mhalachev thanks for your suggestions.  Firstly, I deliberately left out the icons after trying it.  It looked better without, although I do understand your point about teaching new users what the icons mean.  Perhaps something to revisit in another iteration, if we can keep the visual impact down.

Secondly, the Recent Projects and Recent Files sections are similar to what I had in mind for a dashboard to replace the welcome screen.  The background area is only rarely on display - a dashboard accessible from more places could better display those things IMO.

Chris2011 commented:
@mhalachev I also like this and I see it the same way as @neilcsmith-net. It would be like a little wizard/startpage/dashboard. The empty editor is not that often empty. For example when you reopen your projects, you need to close all of them. The dashboard is smth that you can reopen it and also to set the option to show it everytime on startup or whatever. Keep it simple here for the empty editor pane.

mbien commented:
I have some notes on document groups somewhere. One day I might try to make them work a bit better with project groups since right now they are not very useful in the current state (and they are slow) - stashing a doc group would lead right to the empty editor background, so assuming it works we might see that panel more often.

Chris2011 commented:
I guess that the case is that you already have more often document tabs instead of closing all documents or closing a whole group. I will not say, that this case will never happen for someone but how often do you all see the empty editor background in your daily work from using NetBeans every day? I also haven't seen it in vs code. And this thing worked for years for the most IDEs. Yes, we can change things comparing to other IDEs and don't do the exact same, but we should do it not in the wrong way IMHO. The empty editor background should be just for the common stuff for opening the IDE once after new installation, remove the empty editor or in rare other cases.

A dashboard with more stuff like seeing tutorials, some common actions (similar to the empty editor background), recent projets/project groups, etc. is a better use case for a dashboard/startpage that you can always show on startup or before NetBeans starts a little simple dashboard before opening everything in NetBeans to change your case. My 2 cents.

mhalachev commented:
After reviewing again (and playing a little bit more with the dev build), I concur with the afore-mentioned suggestions to maintain simplicity, just the essentials, avoiding excessive dashboard-like elements.

jtulach commented:
This is a very nice visual improvement. Makes NetBeans look more modern. Thank you.

neilcsmith-net commented:
Thanks @jtulach :smile:  Just need to track down a little issue with it not displaying on first open, which I saw while working on something else.

@Chris2011 @mhalachev have opened #7239 which covers my idea for a start on a dashboard to replace the welcome page.

Chris2011 commented:
@neilcsmith-net love it, nice work. Thx :)
